### PR TITLE
VAULT-11510 Vault Agent can start listeners without caching

### PR DESCRIPTION
**Pull request by VioletHynes**
This enables the usage of listeners and the API proxy even without the caching block. Everything works in both cases, and requests to authenticated endpoints (like token lookup) work in both.

I made the decision to make `cache-clear` return a 200 in the case that a cache does not exist, as I think that behaviour probably makes the most sense, but I'm welcome to being challenged on it.

I also updated a lot of the docs around this area. 

There were also config items that were using "caching" to refer to the requests made by the API proxy. I allowed "proxying" as an alternative to make things clearer, hopefully for both us and customers.

### CODE REVIEW

**Comment by VioletHynes** [comment]
Oh, woops, there are merge conflicts. Looking to fix those.

File: command/agent/cache/handler.go [hunk 1]
```diff
@@ -36,7 +90,7 @@ func Handler(ctx context.Context, logger hclog.Logger, proxier Proxier, inmemSin
 		}
 
 		// Parse and reset body.
-		reqBody, err := ioutil.ReadAll(r.Body)
+		reqBody, err := io.ReadAll(r.Body)
```
Comment (VioletHynes) [inline command/agent/cache/handler.go:93]: `ioutil.ReadAll` is deprecated in favour of `io.ReadAll`

File: command/agent/cache/handler.go [hunk 2]
```diff
@@ -45,7 +99,7 @@ func Handler(ctx context.Context, logger hclog.Logger, proxier Proxier, inmemSin
 		if r.Body != nil {
 			r.Body.Close()
 		}
-		r.Body = ioutil.NopCloser(bytes.NewReader(reqBody))
+		r.Body = io.NopCloser(bytes.NewReader(reqBody))
```
Comment (VioletHynes) [inline command/agent/cache/handler.go:102]: `ioutil.NopCloser` is deprecated in favour of `io.NopCloser`

File: website/content/docs/agent/caching/index.mdx [hunk 3]
```diff
@@ -134,7 +134,8 @@ belonging to the cached response, the `request_path` that resulted in the
 cached response, the `lease` that is attached to the cached response, the
 `namespace` to which the cached response belongs to, and a few more. This API
 exposes some factors through which associated cache entries are fetched and
-evicted.
+evicted. For listeners without caching enabled, this API will still be available,
+but will do nothing (there is no cache to clear) and will return a `200` response.
```
Comment (peteski22) [inline website/content/docs/agent/caching/index.mdx:138]: Q: When caching is enabled, does the call to the API endpoint return any content in the response body? 

If **yes**, I think we should adjust the 'disabled caching' response code to `204 (No Content)`.
If **no**, I'm not sure if that's a bit weird semantically but hey ho, I guess it makes no difference if we 'keep it the same' here.

File: command/agent/config/config.go [hunk 4]
```diff
@@ -285,7 +285,9 @@ func LoadConfig(path string) (*Config, error) {
 		case "auto-auth":
 			result.DisableIdleConnsAutoAuth = true
 		case "caching":
-			result.DisableIdleConnsCaching = true
+			result.DisableIdleConnsAPIProxy = true
```
Comment (peteski22) [inline command/agent/config/config.go:288]: Not sure if this is just a style preference, but I recently saw some switch statements where multiple cases had the same block like:

```
case "caching", "proxying": 
    result.DisableIdleConnsAPIProxy = true
```

Not sure if that's something we'd want to follow, or if it makes it easier to miss the 2nd case on the line. Anyway, just worth a mention. Up to you either way is fine by me. 😄

File: command/agent/cache/handler.go [hunk 5]
```diff
@@ -20,7 +20,61 @@ import (
 	"github.com/hashicorp/vault/sdk/logical"
 )
 
-func Handler(ctx context.Context, logger hclog.Logger, proxier Proxier, inmemSink sink.Sink, proxyVaultToken bool) http.Handler {
+func CachelessHandler(ctx context.Context, logger hclog.Logger, proxier Proxier) http.Handler {
```
Comment (peteski22) [inline command/agent/cache/handler.go:23]: This is another firmly 'take it or leave it' comment:

I'm wondering if the two handlers could have naming that matched up slightly better? I'm not 100% sure I know what that is (`CachingHandler` and `NonCachingHandler` maybe)?

File: command/agent/cache/cache_test.go [hunk 6]
```diff
@@ -39,18 +39,26 @@ path "*" {
 // be deferred immediately along with two clients, one for direct cluster
 // communication and another to talk to the caching agent.
 func setupClusterAndAgent(ctx context.Context, t *testing.T, coreConfig *vault.CoreConfig) (func(), *api.Client, *api.Client, *LeaseCache) {
-	return setupClusterAndAgentCommon(ctx, t, coreConfig, false)
+	return setupClusterAndAgentCommon(ctx, t, coreConfig, false, true)
+}
+
+// setupClusterAndAgentNoCache is a helper func used to set up a test cluster and
+// proxyi agent against the active node. It returns a cleanup func that should
```
Comment (peteski22) [inline command/agent/cache/cache_test.go:46]:
```suggestion
// proxying agent against the active node. It returns a cleanup func that should
```

**Review by peteski22 — APPROVED** [review]
Looks good to me, I added some random 'bits' just so you know I read it cover to cover 👍🏼 

I think there's only 1 typo that the suggestion probably needs accepting for, I'll approve it anyway and you can deal with that rather than 'request changes' seems a bit drawn out for a tiny thing.

File: website/content/docs/agent/caching/index.mdx (hunk 3 above)
Comment (VioletHynes) [reply website/content/docs/agent/caching/index.mdx:138]: It doesn't return content in either case (caching or no), but returns a `200`. I'd broadly agree that it 'should' be a `204`, but my goal was to keep the return values the same between the two. If we changed the cache endpoint to `204`, we should update this one too.

File: command/agent/config/config.go (hunk 4 above)
Comment (VioletHynes) [reply command/agent/config/config.go]: Thanks! Adjusted - I think it makes it clearer!

**Comment by peteski22** [comment]
@finnstech: https://github.com/hashicorp/vault/pull/18092 can be closed once this (`18137`) PR is merged.

File: command/agent/cache/handler.go (hunk 5 above)
Comment (ncabatoff) [reply command/agent/cache/handler.go:23]: Or `CachingProxyHandler` vs `ProxyHandler`?

Comment (VioletHynes) [reply command/agent/cache/handler.go:23]: Oh that's a good suggestion! Thanks, I'll go with `CachingProxyHandle` and `ProxyHandler`

File: command/agent/cache/handler.go [hunk 7]
```diff
@@ -20,7 +20,61 @@ import (
 	"github.com/hashicorp/vault/sdk/logical"
 )
 
-func Handler(ctx context.Context, logger hclog.Logger, proxier Proxier, inmemSink sink.Sink, proxyVaultToken bool) http.Handler {
+func ProxyHandler(ctx context.Context, logger hclog.Logger, proxier Proxier) http.Handler {
```
Comment (ncabatoff) [inline command/agent/cache/handler.go:23]: It doesn't look like you included auto-auth support here, how come?  I don't see that as intrinsically tied to the cache.

More generally, would it be possible to factor out more of the common elements between the two handlers?  I don't mind if there's some code duplication, but I think the two handlers are going to be extremely similar.

Comment (VioletHynes) [reply command/agent/cache/handler.go:23]: The auto-auth token is cached today if `config.Cache.UseAutoAuthToken` is set. I don't think it's a good idea to enable this by default, as it could be considered a security issue (any request to Agent could use its auto-auth token). As a result, I'm more inclined to leave this as something users have to opt in to doing (we don't cache the token unless the config is set).

I can definitely try and refactor the handlers. I'll get on that.

Comment (VioletHynes) [reply command/agent/cache/handler.go:23]: I see what you mean - while we do need somewhere to store the token, we don't really need a full cache. I don't really like that we'll be duplicating the config, but I guess that's the price we pay for choosing to implement it this way to begin with. It's doable, like using a new HCL construct like this:

```
api_proxy {
  use_auto_auth_token = true
  force_auto_auth_token = true
}
```

Though this might get confusing as one would have to override the other and we can't deprecate the cache config.

Does this approach make sense?

Comment (ncabatoff) [reply command/agent/cache/handler.go:23]: It's worse than it seems actually, because we want all the settings currently in `cache` except for `persist`, so it's not just the auto auth things.

>  we can't deprecate the cache config.

How do you mean?

**Comment by VioletHynes** [comment]
After a discussion over Slack, I'm going to update this PR to rely upon a new HCL stanza, `api_proxy`, which contains the values from the `cache` stanza relating to api proxying (everything but `persist`). That way, these values can be supported without using a cache.

**Comment by VioletHynes** [comment]
I just pushed an update to support the `api_proxy` stanza - I think most of the code is done, though I'm not sure yet (I'm going to run through a few more manual tests), though basic manual tests seem to work great.

I'm going to write a new docs page about the API Proxy in general, as we don't actually have one.

Lastly, it might be nice to refactor the code so that the `handler` and `api_proxy` stuff isn't in the `cache` package. Doing so naively would result in circular dependencies, so I might look at this as part of a separate PR, once this has been merged.

tl;dr: the PR isn't done yet, and I'll re-request review when it is.

File: command/agent/cache/handler.go (hunk 7 above)
Comment (VioletHynes) [reply command/agent/cache/handler.go]: Resolving this (output of this discussion captured in another comment)

File: command/agent/cache/api_proxy_test.go [hunk 8]
```diff
@@ -93,3 +145,182 @@ func TestAPIProxy_queryParams(t *testing.T) {
...
+	apiProxy, err := NewAPIProxy(&APIProxyConfig{
+		Client: clienToUse,
+		Logger: apiProxyLogger,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create a muxer and add paths relevant for the lease cache layer and API proxy layer
+	mux := http.NewServeMux()
```
Comment (VioletHynes) [inline command/agent/cache/api_proxy_test.go:261]: These functions have been moved over from the `cache_test.go` file, though the next ~25 lines of code were modified on the way.

**Comment by VioletHynes** [comment]
Okay! I think we're good to go for review. Find the new doc page here: https://vault-hffnd61a7-hashicorp.vercel.app/vault/docs/agent/apiproxy

We now support the `api_proxy` stanza and note that the values in the `cache` stanza are deprecated (though they still work, for backwards compatibility). We error if both are provided. Everything I have tested manually seems to work, and all of our tests (that test the proxy and the cache) all seem to pass.

I think I've updated everywhere that needs to be updated, but let me know if I missed anything.

**Review by peteski22 — CHANGES_REQUESTED** [review]
Thanks for the PR Violet, sorry the re-review took me longer than I wanted.

I spent a while going over the docs since I'm newer to Vault Agent and wanted to make sure I understood it. 

I've made a few suggestions which I think we need (typo type stuff), some that I think make the docs clearer (to me) which you can take or leave, and the occasional rambling where I was trying to figure things out.

**Comment by peteski22** [comment]
Random, and not your code/docs etc. but any chance while you're in there you can fix this one thing that for some reason has been making my eye twitch: `docs/agent#configuration`:

> These are the currently-available general configuration option:

to

> These are currently the available general configuration options:

File: command/agent.go [hunk 9]
```diff
@@ -433,9 +461,19 @@ func (c *AgentCommand) Run(args []string) int {
...
+			if whenInconsistent != cache.WhenInconsistentFail {
+				c.UI.Error("enforce_consistency configured in both api_proxy and cache blocks. Please remove this configuration from the cache block.")
+				return 1
+			} else {
+				whenInconsistent = cache.WhenInconsistentRetry
+			}
 		case "forward":
-			whenInconsistent = cache.WhenInconsistentForward
+			if whenInconsistent != cache.WhenInconsistentFail {
+				c.UI.Error("enforce_consistency configured in both api_proxy and cache blocks. Please remove this configuration from the cache block.")
```
[the remark this answers is not in the document]
Comment (VioletHynes) [reply command/agent.go]: Good catch, thanks!

File: website/content/docs/agent/caching/index.mdx [hunk 10]
```diff
@@ -251,14 +223,26 @@ or the default role, `default`, which serves everything (including metrics).
...
 # ...
 
 cache {
-  use_auto_auth_token = true
+	persist = {
+		type = "kubernetes"
+		path = "/vault/agent-cache/"
+		keep_after_import = true
+		exit_on_err = true
+		service_account_token_file = "/tmp/serviceaccount/token"
```
[the remark this answers is not in the document]
Comment (VioletHynes) [reply website/content/docs/agent/caching/index.mdx:239]: Hmmm, good point, this does seem to be missing in our docs. Let me see if I can fix that

[the remark this answers is not in the document]
Comment (VioletHynes) [reply website/content/docs/agent/caching/index.mdx:239]: Ah! It's in the kubernetes section, but I'll add it to the main section too

**Review by peteski22 — APPROVED** [review]
